org-node/src/webserver.rs

```diff
@@ -0,0 +1,170 @@
+    spawn(async move {
+        while let Some(message) = peer_unbounded_rx.next().await {
+            if let Ok(msg) = serde_json::to_string(&message) {
+                peer_ws_tx
+                    .send(Message::text(msg))
+                    .unwrap_or_else(|e| {
+                        eprintln!("websocket send error: {}", e);
+                    })
+                    .await;
+            }
+        }
+    });
+
+    // Send message to update connected web socket clients.
+    if let Some(addr) = remote {
+        if let Err(e) = conn_tx.send((addr, peer_unbounded_tx)) {
+            tracing::error!(
+                target: "org-node",
+                "Failed to inform new websocket client connection: {}",
+                e
+            );
+        }
+
+        // Ignore incoming messages from the connected peer.
+        #[allow(clippy::redundant_pattern_matching)]
+        while let Some(_) = peer_ws_rx.next().await {
+            // Do nothing.
+        }
+    }
+}
+
+/// Serves a warp web server instance with a websocket endpoint for subscribing to events.
+pub async fn serve(listen: std::net::SocketAddr, mut events: UnboundedReceiverStream<WsEvent>) {
+    // Spawn connected clients receiver
+    let (conn_ws_tx, conn_ws_rx) = mpsc::unbounded_channel::<EstablishWebSocketSender>();
+    let mut conn_ws_rx = UnboundedReceiverStream::new(conn_ws_rx);
+
+    // instantiate an empty peer socket map;
+    let connected_ws_clients = ConnectedWebSocketClients::default();
+    let cloned_ws_clients = Arc::clone(&connected_ws_clients);
+
+    // listen for internal events and broadcast to connected peers.
+    tokio::task::spawn(async move {
+        while let Some(msg) = events.next().await {
+            // send update to all connected web socket clients.
+            for (addr, tx) in cloned_ws_clients.read().await.iter() {
+                if let Err(e) = tx.send(msg.clone()) {
+                    tracing::error!(target: "org-node", "Failed to send message to web socket client: {}", e);
+
+                    // Remove disconnected client from connected clients map.
+                    cloned_ws_clients.write().await.remove(addr);
+                }
+            }
+        }
+    });
+
+    // handle adding new connected websocket clients.
+    tokio::task::spawn(async move {
+        while let Some((addr, tx)) = conn_ws_rx.next().await {
+            connected_ws_clients.write().await.insert(addr, tx);
+        }
+    });
+
+    let connected_peers_filter = warp::any().map(move || conn_ws_tx.clone());
+
+    let routes = warp::path(WEBSOCKET_PATH)
+        .and(warp::addr::remote())
+        .and(warp::ws())
+        .and(connected_peers_filter)
+        .map(
+            move |remote: Option<SocketAddr>,
+                  ws: Ws,
+                  tx: mpsc::UnboundedSender<EstablishWebSocketSender>| {
+                ws.on_upgrade(move |socket| establish_connection(socket, remote, tx))
```

Comment:
We should `unwrap` the remote before passing it in, it's only `None` when using a domain socket or something, but since this webserver's listen address is a `SocketAddr`, we know it will have an address.

Reply:
commit 0452b3c fixes.